File: solutions/csharp/perfect-numbers/PerfectNumbers.cs
public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static int AliquotSum(int n)
    {
        if (n < 1) {
            throw new ArgumentOutOfRangeException("");
        }

        int sum = 0;

        for (int i = 1; i <= n-1; i++) {
            sum += n % i == 0 ? i : 0;
        }

        return sum;
    }

    public static Classification Classify(int n)
    {
        if (n == AliquotSum(n)) {
            return Classification.Perfect;
        }
        else if (n < AliquotSum(n)) {
            return Classification.Abundant;
        }
        else {
            return Classification.Deficient;
        }
    }
}

File: solutions/haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

aliquotFactors :: Int -> [Int]
aliquotFactors n = [m | m <- [1..n-1], mod n m == 0]

aliquotSum :: Int -> Int
aliquotSum n = sum (aliquotFactors n)

classify :: Int -> Maybe Classification
classify n
    | n <= 0 = Nothing
    | n == s = Just Perfect
    | n < s = Just Abundant
    | n > s = Just Deficient
    | otherwise = Nothing
  where s = aliquotSum n
